api/lib/opentelemetry/instrumentation/adapter.rb

```diff
@@ -0,0 +1,244 @@
+        def adapter_name(adapter_name = nil)
+          if adapter_name
+            @adapter_name = adapter_name
+          else
+            @adapter_name ||= infer_name || 'unknown'
+          end
+        end
+
+        # Optionally set the version of this adapter. If not explicitly set,
+        # the version will default to the version constant under namespace of
+        # the class, or the version constant under the class name if it does not
+        # have a namespace. If a version constant cannot be found, it defaults
+        # to '0.0.0'.
+        #
+        # @param [String] adapter_version The version of the adapter package
+        def adapter_version(adapter_version = nil)
+          if adapter_version
+            @adapter_version = adapter_version
+          else
+            @adapter_version ||= infer_version || '0.0.0'
+          end
+        end
+
+        # The install block for this adapter. This will be where you install
+        # instrumentation, either by framework hook or applying a monkey patch.
+        #
+        # @param [Callable] blk The install block for this adapter
+        # @yieldparam [Hash] config The adapter config will be yielded to the
+        #   install block
+        def install(&blk)
+          @install_blk = blk
+        end
+
+        # The present block for this adapter. This block is used to detect if
+        # target library is present on the system. Typically this will involve
+        # checking to see if the target gem spec was loaded or if expected
+        # constants from the target library are present.
+        #
+        # @param [Callable] blk The present block for this adapter
+        def present(&blk)
+          @present_blk = blk
+        end
+
+        # The compatible block for this adapter. This check will be run if the
+        # target library is present to determine if it's compatible. It's not
+        # required, but a common use case will be to check to target library
+        # version for compatibility.
+        #
+        # @param [Callable] blk The compatibility block for this adapter
+        def compatible(&blk)
+          @compatible_blk = blk
+        end
+
+        def instance
+          @instance ||= new(adapter_name, adapter_version, install_blk,
+                            present_blk, compatible_blk)
+        end
+
+        private
+
+        attr_reader :install_blk, :present_blk, :compatible_blk
+
+        def infer_name
+          @inferred_name ||= if (md = name.match(NAME_REGEX)) # rubocop:disable Naming/MemoizedInstanceVariableName
+                               md['namespace'] || md['classname']
+                             end
+        end
+
+        def infer_version
+          return unless (inferred_name = infer_name)
+
+          mod = inferred_name.split('::').map(&:to_sym).inject(Object) do |object, const|
+            object.const_get(const)
+          end
+          mod.const_get(:VERSION)
```

Comment:
See? That inferencing is not complicated at all! 🙈 :trollface:

Reply:
😅